### PR TITLE
Asset check severity

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

Introduces `AssetCheckSeverity` for asset checks, which can be set on either an `AssetCheckSpec` or directly when defining an asset check via the `@asset_check` decorator.

By setting its severity level to ERROR, you can specify that a check impacts control flow, i.e. only materialize downstream assets if the check on the upstream asset succeeds.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/dependency.py [hunk 1]
```diff
@@ -724,11 +729,15 @@ def the_job():
     Args:
         dependencies (List[Union[DependencyDefinition, Type[MappedInputPlaceHolder]]]): List of
             upstream dependencies fanned in to this input.
+        dependencies_not_to_load_from (Optional[Sequence[DependencyDefinition]])): A subset of the
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/dependency.py:732]: New attr on `MultiDependencyDefinition`

File: python_modules/dagster/dagster/_core/execution/plan/plan.py [hunk 2]
```diff
@@ -463,29 +463,37 @@ def get_step_input_source(
 
     if dependency_structure.has_fan_in_deps(input_handle):
         sources: List[StepInputSource] = []
+        sources_to_load_from: List[StepInputSource] = []
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/execution/plan/plan.py:466]: Gnarly logic for constructing `FromMultipleSources`

File: python_modules/dagster/dagster/_core/definitions/assets_job.py [hunk 3]
```diff
@@ -450,12 +465,32 @@ def build_node_deps(
...
+                asset_dep_def = DependencyDefinition(upstream_node_alias, upstream_output_name)
+                if blocking_asset_check_output_handles:
+                    deps[node_key][input_name] = MultiDependencyDefinition(
+                        dependencies=asset_check_deps + [asset_dep_def],
+                        dependencies_not_to_load_from=asset_check_deps,
+                    )
+                else:
+                    deps[node_key][input_name] = asset_dep_def
+            elif asset_check_deps:
+                deps[node_key][input_name] = MultiDependencyDefinition(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/assets_job.py:491]: Where we construct the MultiDependencyDefinition

**Review by alangenfeld — COMMENTED** [review]
re-using the existing pieces with these options that change the behavior so much feels really rough. I think it will make future changes much more difficult. 

I think its worth a go at implementing it with dedicated objects for modeling this behavior. I don't think code duplication is worth fretting about (and can be solved in other ways) and the names are internal so we can pick something even if we dont love it

**Comment by sryza** [comment]
@alangenfeld any thoughts on what the name would be for the new kind of `DependencyDefinition`?

**Comment by alangenfeld** [comment]
* `BlockingDependencyDefinition` 
* `DependencyWithBlockingChecksDefinition`
* `DependencyWithBlockingAssetChecksDefinition` 
* `DependencyWithBlocksDefinition`
* `VariantDependencyDefinition` 
* `MultiVariateDependencyDefinition`
* `MergedDependencyDefinition`

**Comment by sryza** [comment]
@alangenfeld I re-spun this by adding a `BlockingAssetChecksDefinition

**Comment by alangenfeld** [comment]
>  Can't solve with inheritance because they're NamedTuples.

what constraint does that have? They already inherit from `StepInputSource`, I would assume you could add an intermediate base class like `MultiStepInputSource` to share code

File: python_modules/dagster/dagster/_core/definitions/dependency.py [hunk 4]
```diff
@@ -768,6 +768,46 @@ def get_dependencies_and_mappings(
         return self.dependencies
 
 
+class BlockingAssetChecksDependencyDefinition(
+    IDependencyDefinition,
+    NamedTuple(
+        "_MultiDependencyDefinition",
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/dependency.py:774]: update str

File: python_modules/dagster/dagster/_core/execution/plan/inputs.py [hunk 5]
```diff
@@ -751,25 +779,44 @@ def load_input_object(
...
-    ) -> Optional[str]:
-        return join_and_hash(
-            *[
-                inner_source.compute_version(step_versions, job_def, resolved_run_config)
-                for inner_source in self.sources
-            ]
+@whitelist_for_serdes
+class FromMultipleSourcesLoadSingleSource(
+    NamedTuple(
+        "_FromMultipleSources",
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/execution/plan/inputs.py:786]: update str (i dont think these matter but feels right have them in sync)

**Comment by sryza** [comment]
> what constraint does that have? They already inherit from StepInputSource, I would assume you could add an intermediate base class like MultiStepInputSource to share code

Ok great point. I just posted an update with a `FromMultipleSourcesLoadSingleSource`

**Review by alangenfeld — APPROVED** [review]
nice, looks crispy clean to me

**Pull request merged**